src/imageview.cpp

```diff
@@ -101,13 +105,71 @@ void ImageView::mouseDoubleClickEvent(QMouseEvent* event) {
 }
 
 void ImageView::mousePressEvent(QMouseEvent * event) {
-  QGraphicsView::mousePressEvent(event);
-  if(cursorTimer_) cursorTimer_->stop();
+  if (currentTool == ToolNone) {
+    QGraphicsView::mousePressEvent(event);
+    if(cursorTimer_) cursorTimer_->stop();
+  } else {
+    startPoint = mapToScene(event->pos()).toPoint();
+  }
 }
 
 void ImageView::mouseReleaseEvent(QMouseEvent* event) {
-  QGraphicsView::mouseReleaseEvent(event);
-  if(cursorTimer_) cursorTimer_->start(CURSOR_HIDE_DELY);
+  if (currentTool == ToolNone) {
+    QGraphicsView::mouseReleaseEvent(event);
+    if(cursorTimer_) cursorTimer_->start(CURSOR_HIDE_DELY);
+  } else if (!image_.isNull()) {
+    QPoint endPoint = mapToScene(event->pos()).toPoint();
+
+    QPainter painter(&image_);
+    painter.setPen(QPen(Qt::red, 5));
+
+    switch (currentTool) {
+    case ToolArrow:
+      drawArrow(painter, startPoint, endPoint, M_PI / 8, 25);
+      break;
+    case ToolRectangle:
+      painter.drawRect(QRect(startPoint, endPoint));
+      break;
+    case ToolCircle:
+      painter.drawEllipse(QRect(startPoint, endPoint));
+      break;
+    case ToolNumber:
+    {
+      // Set the font
+      QFont font;
+      font.setPixelSize(32);
+      painter.setFont(font);
+
+      // Calculate the dimensions of the text
+      QString text = QString("%1").arg(nextNumber++);
+      QRectF textRect = painter.boundingRect(image_.rect(), 0, text);
+      textRect.moveTo(endPoint);
+
+      // Calculate the dimensions of the circle
+      qreal radius = qSqrt(textRect.width() * textRect.width() +
+                           textRect.height() * textRect.height()) / 2;
+      QRectF circleRect(textRect.left() + (textRect.width() / 2 - radius),
+                        textRect.top() + (textRect.height() / 2 - radius),
+                        radius * 2, radius * 2);
+
+      // Draw the path
+      QPainterPath path;
+      path.addEllipse(circleRect);
+      painter.fillPath(path, Qt::red);
+      painter.drawPath(path);
+
+      // Draw the text
+      painter.setPen(Qt::white);
+      painter.drawText(textRect.bottomLeft(), text);
+
```

Comment:
The above line doesn't draw the text centered. Instead, it should be:
```c++
      painter.drawText(textRect, Qt::AlignCenter, text);
```
which will result in centered texts:

![center](https://user-images.githubusercontent.com/981076/47719023-207f0800-dc5f-11e8-8eb0-e10211c619ee.png)